### DISCUSSION TITLE
UTF-8 error in mqttconnection

### DISCUSSION THREAD

**Discussion by Frida854** [Moddable-OpenSource/moddable#1279] [General]
When I run the example: 'mqttconnection', I only get the date once, and then restarts all the time, with 'host: "test.mosquitto.org".
But when I use 'mosquitto' on a Raspberripi P5, I get this error:
'/Users/poulborg/Projects/moddable/examples/network/mqtt/mqttconnection/main.js (37) # Break: fromArrayBuffer: invalid UTF-8!'
on line:
'mqtt.onMessage = (topic, body) => trace(`received "${topic}": ${String.fromArrayBuffer(body)}\n`);'

When I comment out this line, it runs without problems:
'mqtt.publish("moddable/mqtt/example/random", Math.random());'

I use a regular ESP32.

I am using latest update.

**Comment by phoddie** [comment]
Good find. Thank you for the report.

This was broken by recent work on conformance and optimization of `ArrayBuffer.fromString()`. The change accidentally removed the step of coercing the input argument to a string. It was failing when passed a number (the output of `Math.random()`).

The fix is to change this line:

https://github.com/Moddable-OpenSource/moddable/blob/9b9e6a9b478bd11e58fe3d2b3d8b987d8ade9925/xs/sources/xsDataView.c#L501

To

```c
txString c = fxToString(the, mxArgv(0));
```

**Reply by Frida854** [reply]
Tested and ok.

**Reply by phoddie** [reply]
Thanks for the quick check! We'll get this integrated so no one else bumps into it.

**Reply by phoddie** [reply]
Just FYI - the [commit is here](https://github.com/Moddable-OpenSource/moddable/commit/3a7102260afc0e7c94cb811375d4a6a8cd93526b).
